### PR TITLE
Keep a `ByteData` around in `CodedBufferReader` to avoid repeated `ByteData` allocs

### PR DESCRIPTION
**Pull request by osa1**
In Dart, the only way to convert a list of bytes to an `int` or `double` is by using a `ByteData`.

Currently `CodedBufferReader` allocates a new `ByteData` for every fixed-size `int` or `float` decoding.

With this change we now allocate a `ByteData` once initialization and reuse it.

Benchmark results, using the same benchmark reported in #888:

|                              | Before     | After      | Diff                |
|------------------------------|------------|------------|---------------------|
| AOT                          |  26,763 us |  25,305 us | - 1,458 us, - 5.4%  |
| JIT                          |  24,573 us |  25,521 us | +   948 us, + 3.8%  |
| dart2js -O4                  | 212,000 us | 199,300 us | -12,700 us, - 5.9%  |
| dart2wasm --omit-type-checks |  89,750 us |  78,958 us | -10,792 us, -12.0%  |

### CODE REVIEW

File: protobuf/lib/src/protobuf/coded_buffer_reader.dart [hunk 1]
```diff
@@ -13,17 +13,30 @@ class CodedBufferReader {
...
+          int sizeLimit = DEFAULT_SIZE_LIMIT}) =>
+      CodedBufferReader._(
+        buffer is Uint8List ? buffer : Uint8List.fromList(buffer),
+        recursionLimit,
+        sizeLimit,
+      );
+
+  CodedBufferReader._(Uint8List buffer, int recursionLimit, int sizeLimit)
+      : _buffer = buffer,
+        _byteData = ByteData.sublistView(buffer),
```
Comment (mkustermann) [inline protobuf/lib/src/protobuf/coded_buffer_reader.dart:39]: With this we're paying for a `ByteData` allocation even if we don't need it when decoding a proto. There may be many protos that don't have float/double/fixed-ints in them.

Consider doing that lazily:
```
late final ByteData _byteData = ByteData.sublistView(_buffer);
```

(then you may also undo the changes to the constructor)

**Review by mkustermann — APPROVED** [review]
LGTM with comments

File: protobuf/lib/src/protobuf/coded_buffer_reader.dart (hunk 1 above)
Comment (osa1) [reply protobuf/lib/src/protobuf/coded_buffer_reader.dart:39]: Right, but it's one allocation per top-level `fromBuffer` call, nested messages are decoded using the same `CodedBufferReader`.

I'm not sure how expensive the allocation is, but it will be optimizing small messages at the cost of large ones with float etc. fields.

Do you think it's the right trade-off?

Comment (mkustermann) [reply protobuf/lib/src/protobuf/coded_buffer_reader.dart:39]: I was more thinking about dart2js. We know that JS typed data allocations are expensive. Maybe typed data view (like `ByteData`) allocations are not? In that case, feel free to leave like it is.

Comment (osa1) [reply protobuf/lib/src/protobuf/coded_buffer_reader.dart]: I implemented your suggestion. It didn't make significant different in my benchmarks.
